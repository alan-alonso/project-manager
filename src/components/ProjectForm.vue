<template>
  <v-form ref="form" v-model="valid" lazy-validation>
    <v-text-field
      v-model="title"
      id="title"
      autofocus
      clearable
      required
      counter="50"
      hint="Project Title"
      label="Title"
      prepend-inner-icon="title"
      :rules="titleRules"
      :success="!!title"
    />

    <v-textarea
      @input="e => (description = e)"
      id="description"
      clearable
      counter="200"
      hint="Some description text"
      label="Description"
      prepend-inner-icon="comment"
      :rules="descriptionRules"
      :success="!!description"
    />

    <v-text-field
      v-model="people"
      id="people"
      required
      hint="Number of people participating in the project"
      label="Number of Prople"
      min="1"
      prepend-inner-icon="people"
      type="number"
      :rules="peopleRules"
      step="1.0"
      :suffix="people >= 2 ? 'people' : 'person'"
      :success="!!people"
    />

    <v-btn
      @click="onSubmit"
      color="purple darken-1 white--text"
      class="mr-4"
      :disabled="!valid"
    >ADD PROJECT</v-btn>
  </v-form>
</template>

<script lang="ts">
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Vue, Component } from "vue-property-decorator";
import { Action } from "vuex-class";
import { Project, ProjectStatus } from "../models/project";

@Component({
  name: "ProjectForm"
})
export default class ProjectForm extends Vue {
  valid = true;

  title = "";
  titleRules = [
    (v: string) => !!v || "Name is required",
    (v: string) =>
      (v && v.trim().length >= 10) || "Name must have at least 10 characters",
    (v: string) =>
      (v && v.trim().length <= 50) || "Name must be less than 50 characters"
  ];

  description = "";
  descriptionRules = [
    (v: string) => !!v || "Description is required",
    (v: string) =>
      (v && v.trim().length >= 10) ||
      "Description must have at least 10 characters",
    (v: string) =>
      (v && v.trim().length <= 200) ||
      "Description must have at most 200 characters"
  ];

  people = 1;
  peopleRules = [
    (v: string) => !isNaN(Number(v)) || "Number of prople must be a number",
    (v: number) => v >= 1 || "Number of people must be at least 1",
    (v: number) => v <= 5 || "Number of people must be at most 5"
  ];

  @Action("addProject")
  addProject: any;

  /**
   * Validates the form data and adds the project to the list.
   *
   * @public
   */
  onSubmit() {
    if ((this.$refs.form as Vue & { validate: () => boolean }).validate()) {
      const newProject = new Project(
        Math.random().toString(),
        this.title,
        this.description,
        Math.floor(this.people),
        ProjectStatus.ACTIVE
      );
      this.addProject(newProject);
    }
  }
}
</script>
