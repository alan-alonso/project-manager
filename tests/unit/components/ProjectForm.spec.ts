import Vue from "vue";
import Vuetify from "vuetify";

Vue.use(Vuetify);

import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import ProjectForm from "@/components/ProjectForm.vue";

const localVue = createLocalVue();

describe("ProjectForm.vue", () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let vuetify: any;

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  test("Form submit should call the store's addProject action", async () => {
    // arrange
    const mocks = {
      methods: {
        addProject: jest.fn()
      }
    };

    // act
    const wrapper = mount(ProjectForm, {
      localVue,
      vuetify,
      ...mocks
    });

    // assert
    const titleInput = wrapper.find("#title");
    titleInput.trigger("focus");
    await wrapper.vm.$nextTick();
    (titleInput.element as HTMLInputElement).value = "Testing Project";
    titleInput.trigger("input");
    await wrapper.vm.$nextTick();

    const descriptionElement = wrapper.find("#description");
    descriptionElement.trigger("focus");
    await wrapper.vm.$nextTick();
    (descriptionElement.element as HTMLTextAreaElement).value =
      "Description sample";
    descriptionElement.trigger("input");
    await wrapper.vm.$nextTick();

    const button = wrapper.find(".v-btn");
    button.trigger("click");
    expect(mocks.methods.addProject).toHaveBeenCalled();
  });

  test("Project form should be rendered correctly", () => {
    // act
    const wrapper = shallowMount(ProjectForm, {
      localVue,
      vuetify
    });

    // assert
    expect(wrapper.element).toMatchSnapshot();
  });
});
