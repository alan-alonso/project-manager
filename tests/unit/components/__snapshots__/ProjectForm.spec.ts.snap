// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm.vue Project form should be rendered correctly 1`] = `
<v-form-stub
  lazyvalidation="true"
  value="true"
>
  <v-text-field-stub
    autofocus="true"
    backgroundcolor=""
    clearable="true"
    clearicon="$clear"
    counter="50"
    errorcount="1"
    errormessages=""
    hint="Project Title"
    id="title"
    label="Title"
    loaderheight="2"
    messages=""
    prependinnericon="title"
    required=""
    rules="v => !!v || \\"Name is required\\",v => v && v.trim().length >= 10 || \\"Name must have at least 10 characters\\",v => v && v.trim().length <= 50 || \\"Name must be less than 50 characters\\""
    successmessages=""
    type="text"
    value=""
  />
   
  <v-textarea-stub
    backgroundcolor=""
    clearable="true"
    clearicon="$clear"
    counter="200"
    errorcount="1"
    errormessages=""
    hint="Some description text"
    id="description"
    label="Description"
    loaderheight="2"
    messages=""
    prependinnericon="comment"
    rowheight="24"
    rows="5"
    rules="v => !!v || \\"Description is required\\",v => v && v.trim().length >= 10 || \\"Description must have at least 10 characters\\",v => v && v.trim().length <= 200 || \\"Description must have at most 200 characters\\""
    successmessages=""
    type="text"
  />
   
  <v-text-field-stub
    backgroundcolor=""
    clearicon="$clear"
    errorcount="1"
    errormessages=""
    hint="Number of people participating in the project"
    id="people"
    label="Number of Prople"
    loaderheight="2"
    messages=""
    min="1"
    prependinnericon="people"
    required=""
    rules="v => !isNaN(Number(v)) || \\"Number of prople must be a number\\",v => v >= 1 || \\"Number of people must be at least 1\\",v => v <= 5 || \\"Number of people must be at most 5\\""
    step="1.0"
    success="true"
    successmessages=""
    suffix="person"
    type="number"
    value="1"
  />
   
  <v-btn-stub
    activeclass=""
    class="mr-4"
    color="purple darken-1 white--text"
    tag="button"
    type="button"
  >
    ADD PROJECT
  </v-btn-stub>
</v-form-stub>
`;
